ledger/src/blockstore_meta.rs

```diff
@@ -269,6 +303,278 @@ impl ShredIndex {
     pub(crate) fn insert(&mut self, index: u64) {
         self.index.insert(index);
     }
+
+    #[allow(unused)]
+    fn remove(&mut self, index: u64) {
+        self.index.remove(&index);
+    }
+}
+
+const NUM_U64_PER_SHRED_INDEX_V2: usize = (MAX_DATA_SHREDS_PER_SLOT + 63) / 64;
+type ShredIndexV2Inner = [u64; NUM_U64_PER_SHRED_INDEX_V2];
+
+/// A bit array of shred indices, where each u64 represents 64 shred indices.
+///
+/// The current implementation of [`ShredIndexLegacy`] utilizes a [`BTreeSet`] to store
+/// shred indices. While [`BTreeSet`] remains efficient as operations are amortized
+/// over time, the overhead of the B-tree structure becomes significant when frequently
+/// serialized and deserialized. In particular:
+/// - **Tree Traversal**: Serialization requires walking the non-contiguous tree structure.
+/// - **Reconstruction**: Deserialization involves rebuilding the tree in bulk,
+///   including dynamic memory allocations and re-balancing nodes.
+///
+/// In contrast, our bit array implementation provides:
+/// - **Contiguous Memory**: All bits are stored in a contiguous array of u64 words,
+///   allowing direct indexing and efficient memory access patterns.
+/// - **Direct Range Access**: Can load only the specific words that overlap with a
+///   requested range, avoiding unnecessary traversal.
+/// - **Simplified Serialization**: The contiguous memory layout allows for efficient
+///   serialization/deserialization without tree reconstruction.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ShredIndexV2 {
+    index: ShredIndexV2Inner,
+    num_shreds: usize,
+}
+
+impl Default for ShredIndexV2 {
+    fn default() -> Self {
+        Self {
+            index: [0; NUM_U64_PER_SHRED_INDEX_V2],
+            num_shreds: 0,
+        }
+    }
+}
+
+impl Serialize for ShredIndexV2 {
```

Comment:
I am not sure why you need to implement `Serialize` and `Deserialize` manually with `unsafe` code?
Is this a side-effect of using `u64` instead of `u8`?

If you just use `u8` you can simply use `#[serde(with = "serde_bytes")]`.

Reply:
Even then you only need to manually implement (de)serialize for the index field not the whole `ShredIndexV2`. something like:
```rust
#[derive(Serialize, Deserialize)]
pub struct ShredIndexV2 {
    #[serde(with="...")]
    index: ShredIndexV2Inner,
    num_shreds: usize,
}
```
similar to:
https://github.com/anza-xyz/agave/blob/a2fb9d133/ledger/src/blockstore_meta.rs#L74

Reply:
:roll_eyes:  This code is using bincode, and bincode does not use field names.

Reply:
I think that onus is on you.

You are unnecessary making this code more verbose and complicated which adds maintenance burden and makes it more bug-prone should someone change this code again in the future.
You are claiming that you are doing this in name of "efficiency" with an argument about field names which are totally irrelevant with bincode. okay then show the evidence.

Reply:
i suspect that maybeuninit is probably what's saving time here, but i'll profile

Reply:
the difference is the extra memset to init, plus extra memcpy the Derived impl has to do to copy the result of serde_u64_array::deserialize into the final destination array. You can achieve the same perf _manually_ ser/de in place into MaybeUninit, but it's rather silly and tuple is simpler. Tuple seems good.